File: typescript/src/agent/conversation.ts
// =============================================================================
// Agent Layer — Conversation (Multi-Turn)
// =============================================================================

import type { Agent } from "./index.js";
import type { AgentStream } from "./stream.js";
import type { InferAgentOutput, ORInputItem, RunOptions, RunResult, SchemaLike } from "./types.js";

/**
 * A stateful multi-turn conversation with an agent.
 *
 * Tracks the full items history across `.send()` calls so the agent
 * sees prior turns as context. Each send appends the user message and
 * the assistant's response (including any tool calls) to the history.
 *
 * @example
 * ```typescript
 * const agent = new Agent({ name: 'assistant', instructions: 'Be helpful.' });
 * const conversation = agent.conversation();
 *
 * const r1 = await conversation.send('My name is Alice.');
 * const r2 = await conversation.send('What is my name?');
 * // r2.output mentions "Alice" — the agent remembers from the prior turn
 * ```
 */
export class Conversation<S extends SchemaLike | undefined = undefined> {
  private readonly agent: Agent<S>;
  private readonly items: ORInputItem[] = [];

  constructor(agent: Agent<S>) {
    this.agent = agent;
  }

  /**
   * Send a message and run the agent to completion.
   *
   * The user message and the full agent response (including tool call
   * items) are appended to the conversation history.
   */
  async send(input: string, options?: RunOptions): Promise<RunResult<InferAgentOutput<S>>> {
    this.items.push({ type: "message", role: "user", content: input });

    const result = await this.agent.run([...this.items], options);

    // Append assistant response as a message item for future context
    const outputText =
      typeof result.output === "string" ? result.output : JSON.stringify(result.output);

    // Append tool call items if any were made
    for (const tc of result.meta.toolCalls) {
      this.items.push({
        type: "function_call",
        call_id: tc.callId,
        name: tc.name,
        arguments: typeof tc.input === "string" ? tc.input : JSON.stringify(tc.input),
      });
      this.items.push({
        type: "function_call_output",
        call_id: tc.callId,
        output: tc.error ? JSON.stringify({ error: tc.error }) : JSON.stringify(tc.output),
      });
    }

    this.items.push({
      type: "message",
      role: "assistant",
      content: outputText ?? "",
    });

    return result;
  }

  /**
   * Send a message and stream the agent's response.
   *
   * Important: the conversation history is only updated after the stream
   * completes (i.e. after you consume the stream or call `.result()`).
   */
  stream(input: string, options?: RunOptions): AgentStream<InferAgentOutput<S>> {
    this.items.push({ type: "message", role: "user", content: input });

    const stream = this.agent.stream([...this.items], options);

    const originalResult = stream.result.bind(stream);
    stream.result = async () => {
      const result = await originalResult();

      const outputText =
        typeof result.output === "string" ? result.output : JSON.stringify(result.output);

      for (const tc of result.meta.toolCalls) {
        this.items.push({
          type: "function_call",
          call_id: tc.callId,
          name: tc.name,
          arguments: typeof tc.input === "string" ? tc.input : JSON.stringify(tc.input),
        });
        this.items.push({
          type: "function_call_output",
          call_id: tc.callId,
          output: tc.error ? JSON.stringify({ error: tc.error }) : JSON.stringify(tc.output),
        });
      }

      this.items.push({
        type: "message",
        role: "assistant",
        content: outputText ?? "",
      });

      return result;
    };

    return stream;
  }

  /** Get the current conversation history as items. */
  getItems(): readonly ORInputItem[] {
    return this.items;
  }

  /** Clear the conversation history. */
  clear(): void {
    this.items.length = 0;
  }
}

File: python/src/opperai/agent/_conversation.py
"""Agent SDK — Conversation (multi-turn state management)."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from ._serialize import to_json_str, to_text
from ._stream import AgentStream
from ._types import ResultEvent, RunResult

if TYPE_CHECKING:
    from . import Agent


class Conversation:
    """Stateful multi-turn conversation with an agent.

    Tracks full items history across ``.send()`` calls so the agent
    sees prior turns as context.

    Example::

        conv = agent.conversation()
        r1 = await conv.send("My name is Alice.")
        r2 = await conv.send("What's my name?")  # Remembers "Alice"

        conv.clear()  # Reset history
    """

    def __init__(self, agent: Agent) -> None:
        self._agent = agent
        self._items: list[dict[str, Any]] = []

    async def send(self, input: str, **kwargs: Any) -> RunResult:
        """Send a message and get a response."""
        self._items.append({
            "type": "message",
            "role": "user",
            "content": input,
        })

        result = await self._agent.run(list(self._items), **kwargs)

        # Append tool calls and assistant response to history
        for tc in result.meta.tool_calls:
            self._items.append({
                "type": "function_call",
                "call_id": tc.call_id,
                "name": tc.name,
                "arguments": to_json_str(tc.input) if tc.input is not None else "",
            })
            output = (
                to_json_str({"error": tc.error})
                if tc.error
                else to_json_str(tc.output)
            )
            self._items.append({
                "type": "function_call_output",
                "call_id": tc.call_id,
                "output": output,
            })

        # Append assistant output
        self._items.append({
            "type": "message",
            "role": "assistant",
            "content": to_text(result.output),
        })

        return result

    def stream(self, input: str, **kwargs: Any) -> AgentStream:
        """Stream a conversation turn."""
        return AgentStream(self._stream_generator(input, kwargs))

    async def _stream_generator(self, input: str, kwargs: dict[str, Any]) -> Any:
        """Create the streaming generator and update items on completion."""
        self._items.append({
            "type": "message",
            "role": "user",
            "content": input,
        })

        stream = self._agent.stream(list(self._items), **kwargs)
        result: RunResult | None = None

        async for event in stream:
            if isinstance(event, ResultEvent):
                result = RunResult(output=event.output, meta=event.meta)
            yield event

        # Update history after stream completes
        if result:
            for tc in result.meta.tool_calls:
                self._items.append({
                    "type": "function_call",
                    "call_id": tc.call_id,
                    "name": tc.name,
                    "arguments": to_json_str(tc.input) if tc.input is not None else "",
                })
                output = (
                    to_json_str({"error": tc.error})
                    if tc.error
                    else to_json_str(tc.output)
                )
                self._items.append({
                    "type": "function_call_output",
                    "call_id": tc.call_id,
                    "output": output,
                })

            self._items.append({
                "type": "message",
                "role": "assistant",
                "content": to_text(result.output),
            })

    def get_items(self) -> list[dict[str, Any]]:
        """Return a copy of the accumulated items."""
        return list(self._items)

    def clear(self) -> None:
        """Reset the conversation history."""
        self._items.clear()
